src/source/Ice/ConnectionListener.c

```diff
@@ -277,53 +277,62 @@ PVOID connectionListenerReceiveDataRoutine(PVOID arg)
-                        // and get the decrypted data length.
-                        STATUS_SUCCEEDED(socketConnectionReadData(pSocketConnection,
-                                                                  pConnectionListener->pBuffer,
-                                                                  pConnectionListener->bufferLen,
-                                                                  (PUINT32) &readLen))) {
-
-
-
-                        if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_UDP) {
-                            if (srcAddrBuff.ss_family == AF_INET) {
-                                srcAddr.family = KVS_IP_FAMILY_TYPE_IPV4;
-                                pIpv4Addr = (struct sockaddr_in *) &srcAddrBuff;
-                                MEMCPY(srcAddr.address, (PBYTE) &pIpv4Addr->sin_addr, IPV4_ADDRESS_LENGTH);
-                                srcAddr.port = pIpv4Addr->sin_port;
-                            } else if (srcAddrBuff.ss_family == AF_INET6) {
-                                srcAddr.family = KVS_IP_FAMILY_TYPE_IPV6;
-                                pIpv6Addr = (struct sockaddr_in6 *) &srcAddrBuff;
-                                MEMCPY(srcAddr.address, (PBYTE) &pIpv6Addr->sin6_addr, IPV6_ADDRESS_LENGTH);
-                                srcAddr.port = pIpv6Addr->sin6_port;
+                    for(;;) {
+                        readLen = recvfrom(pSocketConnection->localSocket, pConnectionListener->pBuffer, pConnectionListener->bufferLen, 0,
+                                           (struct sockaddr *) &srcAddrBuff, &srcAddrBuffLen);
+                        if (readLen < 0 ) {
+                            if (errno != EWOULDBLOCK) {
+                                DLOGE("recvfrom() failed with errno %s", strerror(errno));
+                            } else {
+                                break;
                             }
-                            pSrcAddr = &srcAddr;
-                        } else {
-                            // srcAddr is ignored in TCP callback handlers
-                            pSrcAddr = NULL;
+
+                        } else if (readLen == 0) {
+                            CHK_STATUS(doubleListRemoveNode(pConnectionListener->connectionList, pCurNode));
                         }
 
-                        pSocketConnection->dataAvailableCallbackFn(pSocketConnection->dataAvailableCallbackCustomData,
-                                                                   pSocketConnection,
-                                                                   pConnectionListener->pBuffer,
-                                                                   (UINT32) readLen,
-                                                                   pSrcAddr,
-                                                                   NULL); // no dest information available right now.
-                    }
+                        if (readLen >= 0 &&
```

Comment:
These if/else statements seem sub-optimal. Is there a need to do any of these in case the callback is not specified? Can we do a set of simple ifs up-front for error case, etc and save data to be called later outside of the if statements after we check whether the callback is specified.